processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -434,8 +434,49 @@ private SubclassMapping createSubclassMapping(SubclassMappingOptions subclassMap
         }
 
         private boolean isAbstractReturnTypeAllowed() {
-            return method.getOptions().getBeanMapping().getSubclassExhaustiveStrategy().isAbstractReturnTypeAllowed()
-                && !method.getOptions().getSubclassMappings().isEmpty();
+            return !method.getOptions().getSubclassMappings().isEmpty()
+                && ( method.getOptions().getBeanMapping().getSubclassExhaustiveStrategy().isAbstractReturnTypeAllowed()
+                    || isCorrectlySealed() );
+        }
+
+        private boolean isCorrectlySealed() {
+            try {
+                Type mappingSourceType = method.getMappingSourceType();
+                return isCorrectlySealed( mappingSourceType );
+            }
+            catch ( Exception e ) {
+                // SEALED not supported.
+                return false;
+            }
+        }
+
+        private boolean isCorrectlySealed(Type mappingSourceType) {
+            if ( mappingSourceType.isSealed() ) {
+                List<? extends TypeMirror> unusedPermittedSubclasses =
+                    new ArrayList<>( mappingSourceType.getPermittedSubclasses() );
+                method.getOptions().getSubclassMappings().forEach( subClassOption -> {
+                    unusedPermittedSubclasses
+                                             .stream()
+                                             .filter(
+                                                 permitted -> ctx
+                                                                 .getTypeUtils()
+                                                                 .isSameType(
+                                                                     permitted,
+                                                                     subClassOption.getSource() ) )
+                                             .findFirst()
+                                             .ifPresent( unusedPermittedSubclasses::remove );
+                } );
```

Comment:
I think that this would be more readable if we write in without streams. Perhaps use an iterator` for removing  from the `unusedPermittedSubclasses`